Sources/SwiftProtobuf/Google_Protobuf_FieldMask+Extensions.swift

```diff
@@ -184,3 +179,201 @@ extension Google_Protobuf_FieldMask: _CustomJSONCodable {
     return "\"" + jsonPaths.joined(separator: ",") + "\""
   }
 }
+
+extension Google_Protobuf_FieldMask {
+
+  /// Initiates a field mask with all fields of the message type.
+  ///
+  /// - Parameter messageType: Message type to get all paths from.
+  public init<M: Message & _ProtoNameProviding>(
+    allFieldsOf messageType: M.Type
+  ) {
+    self = .with { mask in
+      mask.paths = M.allProtoNames
+    }
+  }
+
+  /// Initiates a field mask from some particular field numbers of a message
+  ///
+  /// - Parameters:
+  ///   - messageType: Message type to get all paths from.
+  ///   - fieldNumbers: Field numbers of paths to be included.
+  /// - Returns: Field mask that include paths of corresponding field numbers.
+  public init<M: Message & _ProtoNameProviding>(
+    fieldNumbers: [Int],
+    of messageType: M.Type
+  ) throws {
+    var paths: [String] = []
+    for number in fieldNumbers {
+      guard let name = M.protoName(for: number) else {
+        throw FieldMaskUtilsError.invalidFieldNumber
+      }
+      paths.append(name)
+    }
+    self = .with { mask in
+      mask.paths = paths
+    }
+  }
+}
+
+extension Google_Protobuf_FieldMask {
+
+  /// Adds a path to FieldMask after checking whether the given path is valid.
+  /// This method check-fails if the path is not a valid path for Message type.
+  ///
+  /// - Parameters:
+  ///   - path: Path to be added to FieldMask.
+  ///   - messageType: Message type to check validity.
+  public mutating func addPath<M: Message>(
+    _ path: String,
+    of messageType: M.Type
+  ) throws {
+    guard M.isPathValid(path) else {
+      throw FieldMaskUtilsError.invalidPath
+    }
+    paths.append(path)
+  }
+
+  /// Converts a FieldMask to the canonical form. It will:
+  ///   1. Remove paths that are covered by another path. For example,
+  ///      "foo.bar" is covered by "foo" and will be removed if "foo"
+  ///      is also in the FieldMask.
+  ///   2. Sort all paths in alphabetical order.
+  public var canonical: Google_Protobuf_FieldMask {
+    var mask = Google_Protobuf_FieldMask()
+    let sortedPaths = self.paths.sorted()
+    var set = Set<String>()
+    for path in sortedPaths {
```

Comment:
Is there a more performant version of this?  Once the list is sorted, can you simply check to see if entry N-1 is the prefix of the next entry (with a dot added)?

i.e. -

Once sorted, we'd have something like:

[ "foo", "foo.bar", "foobaz" ]

So then for index 1, you can see "foo." (entry zero with a dot appended) is the prefix of index 1, so index 1 gets dropped.  Then for index 2, "foo." is not a prefix of "foobaz", so index 2 gets kept.

This might reduce things to a single sort and then a linear walk of the list pulling things over.

Reply:
Oh, does this also have to dedup things?  That would be a tweak to just check for the previous also being an exact match.